Add support for publishing React Native Android bridge to Bintray

mokagio opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
As part of an ongoing work to improve the developer experience of the WordPress Android team, we decided to move away from a submodule-based setup for Gutenberg in the Android app in favor of distributing the package as an AAR via [Bintray](http://bintray.com/), with the option to still build locally.

The changes in this PR represent the first step towards that goal: they enable publishing a generated AAR for the package to Bintray.

### Known limitations

- The setup in `build.gradle` expects this repository to be a subfolder of `mobile-gutenber`.
- Eventually, we'll perform the publishing via CI, but till the point above hasn't been addressed, that won't be possible

## How has this been tested?

The easiest way to test this is to checkout this branch in the `gutenberg` submodule of the `mobile-gutenberg` submodule (🙃) in the WordPress Android project.

From there, hack `build.gradle` to force the JS bundle to be generated so that the resulting AAR has all of the required "_stuff_" for the Android app to use.
To do so, [make `isBundleUpToDate()` return `false`](https://github.com/mokagio/gutenberg/pull/1/commits/b3c9517923397279a6ffc13d8d4c05c1fc556ebe#diff-320e7dacc9e9a83ac8128d299319f82cf6dfa2609e0ebbc043d13de33ea92773L200-R204) and hardcode _both_ `buildGutenbergMobileJSBundle` definitions to `true` ([here's one](https://github.com/mokagio/gutenberg/pull/1/commits/8cc19d71c48ea7bafaf9c77c8831a16d1d578ac7#diff-320e7dacc9e9a83ac8128d299319f82cf6dfa2609e0ebbc043d13de33ea92773L2-R7))

After that, `export` your Bintray username and [API key](https://www.jfrog.com/confluence/display/BT/Uploading#Uploading-GettingyourAPIKey) as `BINTRAY_USER` and `BINTRAY_KEY` and run

```
./gradlew assembleRelease bintrayUpload -PbintrayVersion=$(date "+%Y%m%d-%H%M%S")
```

You can obviously use any value you want for `-PbintrayVersion`.

The tasks should succeeds and you should see a new package on your Bintray account matching the version you set.

You should then be able to test this AAR work on the Android app following the same steps as [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13642).

## Types of changes

New feature: allow to publish the React Native Android bridge as an AAR to Bintray.
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested (only manually, unfortunately).
- [ ] ~~My code follows the WordPress code style.~~ <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] ~~My code follows the accessibility standards.~~ <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] ~~My code has proper inline documentation.~~ <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] ~~I've included developer documentation if appropriate.~~ <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] ~~I've updated all React Native files affected by any refactorings/renamings in this PR.~~ <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mokagio committed:
Remove illegal '/' character from RN Android project name

mokagio committed:
Add Bintray Gradle plugin to React Native Android bridge

mokagio committed:
Add "fat" AAR Gradle plugin to React Native Android bridge

This will allow adding Aztec to the Gutenberg AAR to allow clients to
fallback to it.

mokagio committed:
Configure Aztec import in RN Android bridge source or fat AAR build

Notice the addition of a white color resource. This was required by the
fat AAR generation.

Without it, the task would fail as follows:

```
> Task :reBundleAarRelease
[fat-aar]target: /path/to/repo/packages/react-native-bridge/android/build/outputs/aar/@wordpress_react-native-bridge-release.aar

> Task :verifyReleaseResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':verifyReleaseResources'.
> 1 exception was raised by workers:
  com.android.builder.internal.aapt.v2.Aapt2Exception: Android resource linking failed
  /path/to/repo/packages/react-native-bridge/android/build/intermediates/res/merged/release/drawable/ic_check_24px.xml:6: error: resource color/white (aka org.wordpress.mobile.ReactNativeGutenbergBridge:color/white) not found.
  ...
```

mokagio committed:
Add configuration for Gradle to publish RN Android bridge AAR to Bintray

mokagio commented on packages/react-native-bridge/android/build.gradle:
```diff
@@ -341,3 +364,62 @@ If they are changed, the isBundleUpToDate flag is switched to false. That flag i
+        repo = 'maven'
+        name = 'react-native-gutenberg-bridge'
+        licenses = ['GPL-2.0']
+        userOrg = 'wordpress-mobile'
+        vcsUrl = 'https://github.com/wordpress-mobile/gutenberg-mobile.git'
```
In Oguz's and @mchowning original prototype, the question was raised of whether to use WordPress Bintray organization and Gutenberg as the VCS repository. I like this idea because it moves closer to the logic existing without the need to be a submodule of `gutenberg-mobile` itself.

mokagio commented on packages/react-native-bridge/android/settings.gradle:
```diff
@@ -1,4 +1,4 @@
-rootProject.name = '@wordpress/react-native-bridge'
+rootProject.name = '@wordpress_react-native-bridge'
```
The `/` is an illegal character as a build error would otherwise point out.

mokagio commented on packages/react-native-bridge/android/src/main/res/values/colors.xml:
```diff
@@ -2,5 +2,6 @@
 <resources>
 
     <color name="status_bar_color">#006088</color>
+    <color name="white">@android:color/white</color>
```
Without this, the build would fail as follows:

```
> Task :reBundleAarRelease
[fat-aar]target: /path/to/repo/packages/react-native-bridge/android/build/outputs/aar/@wordpress_react-native-bridge-release.aar

> Task :verifyReleaseResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':verifyReleaseResources'.
> 1 exception was raised by workers:
  com.android.builder.internal.aapt.v2.Aapt2Exception: Android resource linking failed
  /path/to/repo/packages/react-native-bridge/android/build/intermediates/res/merged/release/drawable/ic_check_24px.xml:6: error: resource color/white (aka org.wordpress.mobile.ReactNativeGutenbergBridge:color/white) not found.
  ...
```

mchowning commented on packages/react-native-bridge/android/build.gradle:
```diff
@@ -137,7 +143,24 @@ repositories {
+    // WordPress Android while building from source, because that would result
+    // in an unsupported "direct local AAR file dependency". That shouldn't be
+    // a blocker becuase CI runs `assemble` tasks and should always do so
+    // without buildind dependencies from source, for performance and
```
Minor typo: `buildind`

mchowning commented:
I know that right now you're more focused on getting this working, but before we merge this I do think we should also remove all of the setup we have to enable gradle to build and use the js bundle since that won't be needed once we start publishing this the bridge as a ginary (i.e., something loosely along [these lines](https://github.com/mchowning/gutenberg/commit/380600e2be9ee7255f1adc9b23e35c86bbc1bad6)).

mchowning commented on packages/react-native-bridge/android/build.gradle:
```diff
@@ -137,7 +143,24 @@ repositories {
+    // investigate, embed only when making a binary build.
+    //
+    // Note that this also means that we cannot run an `assemble` task for
+    // WordPress Android while building from source, because that would result
```
> we cannot run an `assemble` task for WordPress Android while building from source

This is a nasty little gotcha. I'm guessing this happens only when the unqualified `assemble` task is run (as compared to something like `assembleWasabiDebug` which will not try to directly build the bridge since the bridge project doesn't have a task with that name) because it tries to assemble this dependency. FWIW, just noting that a quick way to get this error is to run `./gradlew :@wordpress_react-native-bridge:assemble` from WPAndroid.

It is interesting how the full error message points to hermes being the problem:

> Direct local .aar file dependencies are not supported when building an AAR. The resulting AAR would be broken because the classes and Android resources from any local .aar file dependencies would not be packaged in the resulting AAR. Previous versions of the Android Gradle Plugin produce broken AARs in this case too (despite not throwing this error). The following direct local .aar file dependencies of the :@wordpress_react-native-bridge project caused this error: /Users/matt/dev/a8c/WordPress-Android/libs/gutenberg-mobile/gutenberg/node_modules/hermes-engine/android/hermes-debug.aar

I wonder if this actually might be calling out the issue that was causing the app to crash because of the missing hermes dependency (and that we "fixed" by manually including the hermes aars in the WPAndroid repo). 🤔 Maybe fixing the hermes issue would also fix this issue (and then we wouldn't have to include the hermes dependencies in the WPAndroid--and keep them up-to-date).

Hopefully, we can figure this out, but if we can't, do you think it would be worthwhile to have the WPAndroid build perform a check and abort early (with a descriptive error message) if someone tries to do a build from source that would assemble this dependency?

mokagio replied:
>  do you think it would be worthwhile to have the WPAndroid build perform a check and abort early (with a descriptive error message) if someone tries to do a build from source that would assemble this dependency?

For sure 👍

mokagio commented:
@mchowning 

> before we merge this I do think we should also remove all of the setup we have to enable gradle to build and use the js bundle since that won't be needed once we start publishing this the bridge

Are you suggesting this to keep the work atomic and do publishing & related cleanup in one go, or are there build performance or quality improvements to be gained with this, too?

Either way, I'm happy to make the cleanup part of this PR 😄

mchowning commented:
> Are you suggesting this to keep the work atomic and do publishing & related cleanup in one go, or are there build performance or quality improvements to be gained with this, too?

It's mostly for code cleanup and code clarity. Since the binary build is entirely replacing the old build-the-js-bundle-and-include-it-as-an-asset type build, all the code for building the js-bundle is essentially dead code. It will be easier to take out that dead code now when we have a good understanding of the build already in our heads as opposed to someone else (or us) trying to do it later when we don't have as much context fresh in our minds. Also, removing that dead code can avoid having some weird, and unwanted, side effects; like what you ran into where the build-from-js-build has code that deletes the `node_modules` folders after the build, which there is no reason to keep doing now that we're using the binary build and could be very confusing (as you found out).

mokagio committed:
Add block-support folder from mobile-gutenberg to RN Android bridge AAR

mokagio committed:
Add unsupported blocks folder from mobile-gutenberg to RN Android bridge AAR

mokagio committed:
Add metro.config.js to React Native bridge package

This is an attempt to make the CI build that should build and upload its
AAR to Bintray not run out of memory and fail.

mokagio committed:
Use only one worker in the React Native bridge for CI

mokagio committed:
Add transformer option to RN bridge Metro config

This is to match the exact solution given by StackOverflow to see if it
helps.

mokagio commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/28216.
